presto-main/src/main/java/com/facebook/presto/execution/scheduler/PartialResultQueryTaskTracker.java

```diff
@@ -0,0 +1,123 @@
+    private final WarningCollector warningCollector;
+    private final long startTime;
+    private final Map<TaskId, RemoteTask> taskIdMap;
+    private final Set<TaskId> completedTaskIds;
+
+    private long maxEndTime;
+    private boolean taskSchedulingCompleted;
+    private boolean eligibleForPartialResults;
+
+    public PartialResultQueryTaskTracker(
+            PartialResultQueryManager partialResultQueryManager,
+            double minCompletionRatioThreshold,
+            double timeMultiplier,
+            WarningCollector warningCollector)
+    {
+        this.partialResultQueryManager = requireNonNull(partialResultQueryManager, "partialResultQueryManager is null");
+        this.minCompletionRatioThreshold = minCompletionRatioThreshold;
+        this.timeMultiplier = timeMultiplier;
+        this.warningCollector = requireNonNull(warningCollector, "warningCollector is null");
+        this.startTime = System.nanoTime();
+        this.taskIdMap = new HashMap<>();
+        this.completedTaskIds = new HashSet<>();
+    }
+
+    private double getTaskCompletionRatio()
+    {
+        return (double) completedTaskIds.size() / (double) taskIdMap.size();
+    }
+
+    public void trackTask(RemoteTask task)
+    {
+        taskIdMap.put(task.getTaskId(), task);
+    }
+
+    @Override
+    public void stateChanged(TaskInfo taskInfo)
+    {
+        completedTaskIds.add(taskInfo.getTaskId());
+        if (taskSchedulingCompleted && !eligibleForPartialResults) {
+            if (getTaskCompletionRatio() >= minCompletionRatioThreshold) {
+                // Completion ratio greater than equal to threshold. Query eligible for partial results
+                eligibleForPartialResults = true;
+
+                // Set max task time = timeMultiplier x time taken to reach minCompletionRatioThreshold
+                long elapsedTime = System.nanoTime() - startTime;
+                maxEndTime = startTime + (long) (timeMultiplier * elapsedTime);
+
+                partialResultQueryManager.addQueryTaskTracker(this);
+            }
+        }
+    }
+
+    public long getMaxEndTime()
+    {
+        return maxEndTime;
+    }
+
+    public void setTaskSchedulingCompleted()
+    {
+        this.taskSchedulingCompleted = true;
+    }
+
+    @Override
+    public int compareTo(PartialResultQueryTaskTracker o)
+    {
+        return Long.compare(maxEndTime, o.getMaxEndTime());
+    }
+
+    public void cancelUnfinishedTasks()
+    {
+        SetView<TaskId> pendingTaskIds = difference(taskIdMap.keySet(), completedTaskIds);
+        double partialResultPercentage = getTaskCompletionRatio() * 100;
+        for (TaskId taskId : pendingTaskIds) {
+            RemoteTask pendingTask = taskIdMap.get(taskId);
+
+            // Cancel pending tasks
+            pendingTask.cancel();
+
+            String warningMessage = format("Partial results are returned. Only %.2f percent of the data is read.", partialResultPercentage);
```

Comment:
This should be outside of the for-loop I think? Also, as we are moving it out, make sure we first check if there is any pending tasks before adding warning.